### ISSUE TITLE
Visual stuttering and frame drops regardless of settings

### ISSUE THREAD

**Issue by m-orsh** [mgba-emu/mgba#2199]
Lowered audio buffer, tried 60fps, 59.7275fps, different renderers, everything I could attempt.

There is a consistent frame being dropped every second or two. (Yes frame skip is at 0) I have the audio buffer at the lowest setting, and sync to audio enabled.

I recorded a video to try and upload the stuttering, but amusingly enough, the video recorded at 59.7275fps has zero stuttering when I viewed it.

GTX 1080
i7-9700K

**Comment by RafaelTrepaUnCarballo** [comment]
I have similar issues only in Windows, while in Switch the framerrate and audio are silky smooth. The issue was present in every build I tested (dev builds, stable 0.9.1-0.7.3).

I could "fix" it by enabling **video sync**, disabling **audio sync** and enforcing **Adaptative sync (half refresh rate)** in mgba profile in Nvidia drivers. The downside is the audio sounds terrible (even with the highest audio buffer) and I can't fastforward beyond the refresh rate (120hz, so x2) of my laptop (I can, but it messes the video/audio flow and doesn't really go faster)

**I think FPS target is buggy and raises the framerrate slightly higher than the chosen target with audio sync disabled.**
When I choose 60fps or native, mgba FPS counter show 65-66fps. 45fps option gives 47fps, 30fps option gives 30.6fps, 90fps gives 92-94fps... only 120 seems to run at that framerate and smooth (although 2 times faster), but that's probably due to the windows compositor limiting it to the monitor max refresh rate.

Current Device: Laptop legion-5-15arh05h with ryzen 5 4800H, GTX2060 (and a Radeon disabled in BIOS), 16GB RAM, Win10 20h2 and a **120hz** monitor (some emulator dislike being run at that refresh rate, forcing me to use **Adaptative sync (half refresh rate)**

I also have some trouble with framerrate in my old computer, affecting this and other emulators. In that case, I think the root problem was the dual GPU (intel HD4600+ Geforce 820M) managed by Nvidia Optimus, giving weird framepacing and diagonal tearing issues in many emulators and games. I think I "solved" that by enforcing regular vsync in Nvidia global settings (breaking the turbo in several emus) and running mgba with Intel igpu, but I'm not sure...

**Comment by endrift** [comment]
That issue sounds like #827, which I've yet to be able to reproduce even on my 144 Hz monitor.

**Comment by RafaelTrepaUnCarballo** [comment]
Why can't video sync fill the extra refresh cycles with duplicated frames? Treating refresh rates that are multiple of 60 as 60hz?

**Comment by endrift** [comment]
That's not how drawing works here. If I don't push a frame to the GPU it doesn't update the GPU. And I'm trying not to push extra frames. Moreover, the point of sync to video is to use the monitor's refresh rate to clock the sync. If the monitor's refresh rate is wildly different from the GBA, then it's meaningless for clocking sync and you'd be better off just using a random software timer. Audio sync uses the audio driver to clock the sync, which is actually predicable and easy to adjust against.

**Comment by endrift** [comment]
That's not how drawing works here. If I don't push a frame to the GPU it doesn't update the GPU. And I'm trying not to push extra frames. Moreover, the point of sync to video is to use the monitor's refresh rate to clock the sync. If the monitor's refresh rate is wildly different from the GBA, then it's meaningless for clocking sync and you'd be better off just using a random software timer. Audio sync uses the audio driver to clock the sync, which is actually predicable and easy to adjust against.

**Comment by endrift** [comment]
Does anyone here have this issue who isn't using a high framerate monitor?

**Comment by m-orsh** [comment]
> Does anyone here have this issue who isn't using a high framerate monitor?

To be clear, this original issue occurs on whatever monitor I use it on. 60fps, 120fps, etc. I've got two monitors attached running at 60fps currently. My TV is one of them, and it can't do anything higher than 60fps. Both show this issue.

It occurs regardless of what audio buffer is set.
It occurs regardless of what game is running.
It occurs regardless of whether sync to video or sync to audio is chosen.
It occurs on the latest version of mgba.
It occurs while running at 60fps OR 59.7275 (original) fps.

Here's a video i recorded on OBS.

https://s1.webmshare.com/x8gxn.webm as you can see there are little jumps and hops in the framerate. This is what I'm talking about. (Ignore the one moment i hit the speed button by accident)

Not sure what's up. But it's probably related to something related to PC configuration rather than screen. 
Windows 10 x64 by the way.

**Comment by endrift** [comment]
I can see a few little bumps, yeah. Not very many though, and I'm not sure exactly what's going on there.

**Comment by m-orsh** [comment]
This is the only issue that prevents mgba from looking like original hardware when running for me.

**Comment by TeoTotoro** [comment]
Latest dev version improved it for me. Try it out!

**Comment by Kurayami7** [comment]
I don't mean to nag but is this getting fixed any time soon? I'm prone to headaches and this issue makes the emulator completely unusable for me. It happens in every top-down game I try.

**Comment by endrift** [comment]
It might have been improved recently. Please try a development build. The biggest thing holding it back from being fixed is, well, I can't reproduce it so I can't tell if it's fixed or not without other people telling me.

**Comment by m-orsh** [comment]
Just tried the latest development build. It still stutters, here is some additional info while on default settings opening the emulator:
Audio buffer set to default (1536) -> constant stuttering, very choppy.
Audio buffer set to lowest setting -> least stuttering, but still has a stutter every 1-2 seconds.

**Comment by endrift** [comment]
Hmm, I think my sound card may be affecting the situation some if that's what your results are (they're much worse on my end on Windows at the lowest setting, which is bad). What sound card do you have?

**Comment by m-orsh** [comment]
To be clear when I say lowest setting I mean the 512 setting.
Motherboard is MAG Z390 TOMAHAWK and sound output I believe is Realtek® ALC892 Codec. I don't have any specific PCIE sound card, Realtek HD audio is what drives audio on the motherboard seemingly.

**Comment by endrift** [comment]
Yes, that's the info I was looking for. Do you have any of the extra Realtek software installed? I do, and it's kind of terrible, but I need it for some stuff...

**Comment by m-orsh** [comment]
There's a "Realtek Audio Console" that brings up this when opened:
![image](https://user-images.githubusercontent.com/7002881/123199648-49421480-d47d-11eb-8274-001db3fd9b95.png)

**Comment by endrift** [comment]
Interesting. I'll see what I can do.

**Comment by RafaelTrepaUnCarballo** [comment]
I think the new builds improved a little the frame times in my laptop, but it still not perfect.

Unlike @m-orsh if I reduce much the audio buffer (512) the choppiness increases and the fps counter show values lower than 60fps (around 55ish). The medium or highest values work better for me.

I have a Realtek Audio chip, but I don't know where to find its full name.
My laptop is **Legion 5-15ARH05H Laptop (Lenovo) - Type 82B1** 
Hardware id of Realtek: **HDAUDIO\FUNC_01&VEN_10EC&DEV_0257&SUBSYS_17AA389E&REV_1000**
Windows 10 21H1

**Comment by RafaelTrepaUnCarballo** [comment]
I forgot to mention the issue disappears once I connect my laptop to my 60hz TV.

**Comment by Kurayami7** [comment]
Strange. I have a 60hz monitor and get the issue. 

My other specs 

![Capture](https://user-images.githubusercontent.com/77198254/123435493-f7ca8e80-d610-11eb-9f94-2cfccc926ed1.PNG)

**Comment by RafaelTrepaUnCarballo** [comment]
![imagen](https://user-images.githubusercontent.com/16591467/123457971-c8059180-d5e4-11eb-9c67-935958aa2452.png)
![imagen](https://user-images.githubusercontent.com/16591467/123458015-d358bd00-d5e4-11eb-86ee-a4630e5976d8.png)
![imagen](https://user-images.githubusercontent.com/16591467/123458864-d0aa9780-d5e5-11eb-80b9-dfc702c1c7a4.png)


Another detail I should have mentioned: When I connected it to my TV, I felt some minimal hitches with the largest audio buffer that disappeared when I lowered to lower values (512, 700... I didn't test more).

Curiously, I suffer the laptop shows the opposite effect with its own 120hz monitor: 768 and specially 512 causes extra choppiness and framerate drops down to 56fps.

**Comment by endrift** [comment]
Can you highlight the audio devices on these? I think those are contributing more than you might expect.

I'd like to fix this for 0.9.2 but given that I can't reproduce it locally (as it varies so wildly between computers) I'm not hopeful.

**Comment by RafaelTrepaUnCarballo** [comment]
I'm afraid I can't give more info of the sound card. Windows and Speccy identify it with the generic name "Realtek High Definition Audio", I don't know how to get more information of this hardware.

The Hardware id of the Realktek is **HDAUDIO\FUNC_01&VEN_10EC&DEV_0257&SUBSYS_17AA389E&REV_1000** and it's in the motherboard **LENOVO LNVNB161216 FP6**.

**Comment by endrift** [comment]
Googling around makes me think it's a Realtek ALC3287, which is what I was looking for. I have no idea if that'll be helpful though. I'm still trying to figure out how to reproduce, but I'm not sure I want to hold up 0.9.2 on this.

**Comment by Kurayami7** [comment]
Trying to play Zero Mission and I think the issue is causing intermittent audio crackling. Guess it's not just top down games.

**Comment by RafaelTrepaUnCarballo** [comment]
> Trying to play Zero Mission and I think the issue is causing intermittent audio crackling. Guess it's not just top down games.

Before jumping to conclusions...
-Do you have **Sync to Audio** enabled and **Sync to Video** disabled? (in the Emulation menu)
-Did you try increasing the audio buffer to higher values? (Tools->Settings...)

I ask because I don't suffer "intermittent audio crackling", only some frame shuttering.

**Comment by endrift** [comment]
Audio crackling is an unrelated issue to this one--this one only affects visual presentation.

**Comment by EduardoJordao** [comment]
I'm having this issue on a freesync enabled 75hz monitor. On older Windows 10 revisions and AMD drivers the game displayed smoothly in fullscreen mode like it does on a gba screen, but now on latest versions I get visible shimmering on both window and fullscreen modes (always at native gba refresh rate). I tested older mgba versions, but the issue persisted.

This issue can be mitigated by setting the monitor to 60hz, like I had to do in Sonic 3 AIR, then both window and fullscreen modes behave correctly. The emulator core in Retroarch behaves correctly on 75hz (I have both "Sync to exact content framerate" and "Disable desktop composition" enabled).

Another thing that I found interesting on 0.9.1 openGL renderer is that the fast forward (held) is limiting itself to the display refresh rate (it even ignores AMD settings if I set vsync to Always Off). Amusingly, the content displays correctly that way (except for the audio, which sounds off at 74.9 fps).

[syncIssueReport.zip](https://github.com/mgba-emu/mgba/files/6758638/syncIssueReport.zip)

**Comment by endrift** [comment]
@EduardoJordao That issue is #827, which I think is a different issue.

**Comment by m-orsh** [comment]
It would be the best thing ever for this issue to be fixed, as having a GBA emulator run smoothly at 60FPS (or the original framerate) on my computer would be great

**Comment by RafaelTrepaUnCarballo** [comment]
The latest update of Retroarch's mgba core fixed frame pacing and audio cracking issues.
https://www.libretro.com/index.php/retroarch-1-9-13-release/

I was wondering if the fixed issues of RA core and the active issues of standalone mgba are related...

**Comment by m-orsh** [comment]
This framerate judder is the biggest annoyance I have with mgba, without it it'd be perfect.

**Comment by iamthelaw2041** [comment]
@endrift Can you fix this please? It hurts my eyes and makes me feel sick.

**Comment by endrift** [comment]
Asking nicely doesn't help it get fixed. The reason it hasn't been fixed is because I haven't figured out the source of the issue, not because I don't want to fix it. I assure you, I do want to fix it.

**Comment by Rings4** [comment]
Chiming in that I, too, get stutter in mGBA.

Is there a GBA emulator for Windows that DOESN'T have stutter?

**Comment by GaussTek** [comment]
Hello! I just wanted to report that I also get stuttering, it's more notable on vertical scrolling games and so on, like Kirby Tilt n Tumble. I do manage to fix it by decreasing the Audio Buffer from the default to the lowest setting. Increasing it just triggers heavy stuterring in my case. I do have Audio sync enabled and I have a 144Hz monitor on my Laptop, if that helps.

**Comment by mcmull500** [comment]
I would like to add that I have a 60Hz monitor and I get stuttering in the same way. Audio Buffer helps, but doesn't completely eradicate it.

I would like to point out that on version 0.8.3 I do NOT get stuttering at all. I am upgrading to solve a different issue. but 0.8.3 does not have stuttering for me at all

**Comment by SkyPirate1** [comment]
Here's my 2 cents:

Running a laptop with 60hz monitor and Windows 8.1 64 bit.  I used "Classic NES Series - Castlevania (USA)" rom demo mode as a test because it works really well for this particular case.

-Audio sync is on, sync to video is off, framerate is set to native, frameskip is zero, using OpenGL display driver.
-I tried all combinations of audio buffer plus output frequency.
-Increasing buffer size makes the framerate worse (at 4096 it looks like frameskipping is turned on to maximum even though it is not), and muting the volume makes it the absolute worst surprisingly.

I noticed that in my task manager while this is going on, the background process "Windows Audio Device Graph Isolation" (audiodg.exe) is spiking at around 25% for CPU usage during the worst framerates, and with the volume muted in the emulator itself it hovers around that mark the whole time Castlevania is running its demo.  The title screen spikes are consistent, even though there is no sound played at the title screen until you press start.  The title screen spike is also the same with just the "sync to video" option enabled, yet there is no frame dropping at all with this on.  With "sync to audio" the demo framerate is all over the place, as low as under 50 fps!

Hopefully that all wasn't to hard to follow along to.

Now I am going to tell you what caused this problem.  The built in Windows sound enhancements!!!  Here's what fixed it for me:

Go to control panel and find "Sound" or whatever it is that controls sound in current versions of Windows.  Click it, then in "Speakers properties" go to "Enhancements" tab, and then tick the box "Disable all enhancements" and apply, ok, yada yada.

Odd thing is I had all of those turned off, but the box wasn't ticked.  I guess Windows was still running them regardless.

As a final test I ran "Famicom Mini 29 - Akumajou Dracula (Japan)" because the title is on a loop and scrolls in smoothly from the left.  There is still some slight stutter but it works a lot better than it did.  Framerate is always between 59.5 and 59.7 now.

**Comment by Rings4** [comment]
> Here's my 2 cents:
> 
> Running a laptop with 60hz monitor and Windows 8.1 64 bit. I used "Classic NES Series - Castlevania (USA)" rom demo mode as a test because it works really well for this particular case.
> 
> -Audio sync is on, sync to video is off, framerate is set to native, frameskip is zero, using OpenGL display driver. -I tried all combinations of audio buffer plus output frequency. -Increasing buffer size makes the framerate worse (at 4096 it looks like frameskipping is turned on to maximum even though it is not), and muting the volume makes it the absolute worst surprisingly.
> 
> I noticed that in my task manager while this is going on, the background process "Windows Audio Device Graph Isolation" (audiodg.exe) is spiking at around 25% for CPU usage during the worst framerates, and with the volume muted in the emulator itself it hovers around that mark the whole time Castlevania is running its demo. The title screen spikes are consistent, even though there is no sound played at the title screen until you press start. The title screen spike is also the same with just the "sync to video" option enabled, yet there is no frame dropping at all with this on. With "sync to audio" the demo framerate is all over the place, as low as under 50 fps!
> 
> Hopefully that all wasn't to hard to follow along to.
> 
> Now I am going to tell you what caused this problem. The built in Windows sound enhancements!!! Here's what fixed it for me:
> 
> Go to control panel and find "Sound" or whatever it is that controls sound in current versions of Windows. Click it, then in "Speakers properties" go to "Enhancements" tab, and then tick the box "Disable all enhancements" and apply, ok, yada yada. https://imgur.com/l5AeM7b
> 
> Odd thing is I had all of those turned off, but the box wasn't ticked. I guess Windows was still running them regardless. But the spiking is gone and everything is more silky smooth.
> 
> As a final test I ran "Famicom Mini 29 - Akumajou Dracula (Japan)" because the title is on a loop and scrolls in smoothly from the left. There is still some slight stutter but it works a lot better than it did. Framerate is always between 59.5 and 59.7 now.

Unfortunately disabling audio enhancements is not feasible for some setups . I require it for my surround sound setup, otherwise my pc will just send a stereo signal and not activate my subwoofers.

**Comment by SkyPirate1** [comment]
Right I understand that.  I'm hoping maybe that info might provide some insight on the stuttering problem.

I never really thought much about the cpu usage that audiodg.exe was using before, just thought it was normal and it doesn't affect the other emulators I use such as MAME, snes9x, nestopia, punes, Bizhawk, or even mednafen in its psx core which is a little taxing for this laptop.  I also wonder why it spikes more when there is no sound.  My reasoning is, if there is no sound playing, then there is little to process so the cpu usage should be low.

60 fps setting in mGBA is perfect.  I tried turning the framerate down little by little towards the native rate (59.9, 59.8, etc) and it stutters no matter how rounded the numbers are.

**Comment by m-orsh** [comment]
Since creating the issue I've had sound enhancements disabled.

**Comment by HoratioSanchez** [comment]
Did anyone lese have a solution to this? I switched to 0.8.3 win64-bit and it seemed to stop it.

**Comment by zipplet** [comment]
I'm having this issue with the latest release :(
60hz monitor, Windows 10. Happy to provide more information if it helps nail down this problem.
I've tried all the suggestions here and I do confirm that lowering the audio buffer to 512 improves it slightly, but it's still very bad.

**Comment by DJCPT** [comment]
I too am having this stutter issue. Even when I record at 60FPS and move the file to my phone the stutter is blatantly there meanwhile emulating the same ROM on my Android device runs butter smooth. I have a great computer and using a 170Hz 1440p Acer display. Running at 60Hz does not solve the issue. Changing FPS locks, freesync on/off, and VSYNC on/off does not solve the issue.  I tried 0.8.3 and noticed improvement but it still has the judder/stutter. 

https://user-images.githubusercontent.com/105520185/190065298-1fa59c85-597f-46ad-97b1-2539b9e464a2.mp4

**Comment by endrift** [comment]
Chiming in "me too" is not useful. It's a known issue, one I can reproduce, and one which I've yet to be able to fix despite attempts. Further replies that don't give any new information will be removed.

**Comment by endrift** [comment]
I attempted to fix this for 0.10.0, but ran into enough trouble that I've delayed it until 0.10.1 in the interest of finally releasing 0.10.0.

**Comment by AcerUp** [comment]
This stuttering problem is definitely present on my end in 0.10.1 and all development builds since then, but checking audio sync only works fine for the previous release (0.10.0).

**Comment by AcerUp** [comment]
I found something: adjusting the "Settings - High-resolution scale" to very high values (11x or above) causes the stuttering on newer versions than 0.10.0. Doesn't stutter at all on lower values than 9x, at least from what I can see. Interesting that scaling to 11x or above doesn't have that problem in the previous release, but that should help narrow down the culprit.

**Comment by donileo** [comment]
Seeing stuttering on macOS 13.2.1 with mGBA 0.10.1. For some reason OpenEmu 2.3.3 which uses mGBA 0.10.0 is much smoother with very little stutter.

**Comment by AcerUp** [comment]
mgba development version 7947-ea345ca81 is working a lot better. Still barely noticeable, but 95% of the stuttering is gone, even on 16x high resolution scale. Game tested was super mario advance 4: hammer suit, chucking hammers, and sprinting at full speed through the first level at the same time for reference.

**Comment by RafaelTrepaUnCarballo** [comment]
After updating to the latest dev (0.11-7952-9b2ba64bc), the stuttering and frame drops seem to have disappear!
With Audio sync and a sample rate 1024-1536 it feels perfect

(I also recently updated my Nvidia drivers to 528.49, maybe that helped)

**Comment by clavignejr** [comment]
I just wanted to "chime in," not sure if my comment is going to get removed or not, though, I have the following:
i5 13600k
RTX 3070 Ti

The monitors I use are NZXT QHD 1440p 27-inch monitors with a refresh rate that goes up to 165Hz.

The latest stable along with the latest nightly both recreate the "stuttering issue" for me.
I've tried limiting my monitors refresh rate to 60, same issue. The "sync video" setting fixes it 98%, but stuttering still occurs and the audio crackles no matter the buffer rate. I've tried different display drivers, sample rate, and have messed around with a bunch of random settings, nothing will fix the issue.

What is CRAZY to me, is my work computer that has just an i5 9600K, 16GB DDR4, and an SSD, no dedicated graphics card, runs mGBA better than my gaming setup.

The monitor used at work is a 60Hz 27-inch Dell monitor, and using the default settings within mGBA (sync video disabled, etc.) it runs smooth as butter.

I can't explain it, I know that probably doesn't help, though, it is what I've experienced. I stopped emulating GBA games on computer as a whole due to not really getting a satisfying emulation.

**Comment by socketwrench54** [comment]
I had stuttering, but only with vertical scrolling e.g. moving up and down in Pokémon Emerald. Dropping the audio buffer to 512 samples significantly improved this and is now not noticeable. All other settings are default. Version 0.10.1, last updated today (11/3/23). I tried changing my display from 144 Hz to lower refresh rates but this did not work. I'm back to using 144 Hz without issue. Windows 11.

I'm not sure why this worked, but hopefully it can provide some guidance. Keep up the great work!

**Comment by christianbittner** [comment]
I solved this issue by using the latest dev build and switching to the built-in software solution in "enhacements" and using SDL as "audio driver" with 44.1Khz and 4096 buffer size using both audio and video sync.
This however introduces some slight audio crackling, but its minimized. This might occur bec of some sync issues on the current build. Framerate caps dont seem to work properly.
Its a little sad as we now cant take much advantage of better hardware but since emulation is done mostly by the cpu anyways its good to have that working atleast.
Hope this helps you out for the time being!

**Comment by endrift** [comment]
4096 audio buffers will lead to a noticeable lag, either audio lag if you have sync to video on or video lag (effectively input lag) if you have sync to audio on.

**Comment by christianbittner** [comment]
You are correct. I kinda tinkered around with the latest build for the entire evening and you made me look into it again. I got the best results regarding audio and stable framerate by going for QT audio 2048 buffer and 44.1Khz sample rate. Display driver is OpenGL and Enhancements are done in software (OpenGL introduced tearing of some sorts). Native refresh with sync to audio only.
This however introduced a bug for some reason. Both speed up and autofire seem to behave as toggles now, no matter which option is selected and I cannot deactivate them by using my controller.
This might be because I'm using a controller with analog triggers as using Tab or Shift+Tab fixes the speed up behavior and works as intended.
Is there a workaround?
It properly worked this afternoon and now has issues, turning on speed up AFTER I release the trigger and only goes back to normal speed when I hold the trigger down again, which is the wrong way round. Maybe its relating to the + and - values it gets from the analog trigger idk, but it worked normally some hours ago

**Comment by xXRaptorSc0pezXx** [comment]
**My 2 cents on this issue:**
I just downloaded this emulator and was disappointed it had this issue because it is otherwise great so hopefully my findings can help to solve it.
PC Specs: RTX 3080 12GB, Intel i7-12700k, using a 165hz monitor. I used Stable release 0.10.1 for this testing.

**This stuttering bug occurs:**
- Much more often when the application is fullscreened. Even if I have the actual game's screen size the exact same size when windowed as when fullscreened, fullscreen always caused this bug to occur more frequently.
- ONLY when using Display driver OpenGL. When using OpenGL (force version 1.x) the bug does NOT occur at all. I did not test Software.
- Regardless of Frameskip setting (although higher values cause the stuttering to occur significantly more)
- Regardless of Audio buffer sample setting or Sample rate
- Regardless of Aspect ratio, Integer scaling, Interframe blending, or Bilinear filtering
- Regardless of FPS target
- Regardless of High-resolution scale setting

So, if anyone reading this is having this issue, a good temporary fix until the stuttering bug is fully fixed is to change the Display driver to "OpenGL (force version 1.x)" since for some reason that completely gets rid of the stuttering.

Hope this helps! :D

**Comment by rafaelsetti** [comment]
0.9.1 version works fine here, after this version all the subsequent versions have this same problem here

**Comment by endrift** [comment]
> Chiming in "me too" is not useful. It's a known issue, one I can reproduce, and one which I've yet to be able to fix despite attempts. Further replies that don't give any new information will be removed.

This is still the case. I will be locking the issue until it's fixed.
